inst/stan/functions/discretised_logit_hazard.stan

```diff
@@ -0,0 +1,89 @@
+real loglogistic_lcdf (real y, real alpha, real beta) {
+  return -log1p((y / alpha) ^-beta);
+}
+  
+
+// Compute the cdf of a parametric distribution at values 1:n
+vector upper_lcdf_discrete(real mu, real sigma, int n, int dist) {
+  vector[n] upper_lcdf;
+  if (dist == 1) {
+    real emu = exp(-mu);
+    for (i in 1:n) {
+      upper_lcdf[i] = exponential_lcdf(i | emu);
+    }
+  } else if (dist == 2) {
+    for (i in 1:n) {
+      upper_lcdf[i] = lognormal_lcdf(i | mu, sigma);
+    }
+  } else if (dist == 3) {
+    real emu = exp(mu);
+    for (i in 1:n) {
+      upper_lcdf[i] = gamma_lcdf(i | emu, sigma);
+    }
+  } else if (dist == 4) {
+    real emu = exp(mu);
+    for (i in 1:n) {
+      upper_lcdf[i] = loglogistic_lcdf(i | emu, sigma);
+    }
+  } else {
+    reject("Unknown distribution function provided.");
+  }
+  return(upper_lcdf);
+}
+
+// Adjust a vector of cdf evaluations for probability mass beyond maximum value
+vector adjust_lcdf_discrete(vector lcdf, int n, int max_strat) {
+  vector[n] adjusted_lcdf;
+  if (max_strat == 0) {
+    // ignore
+    adjusted_lcdf = lcdf;
+  } else if (max_strat == 1) {
+    // add to maximum value
+    adjusted_lcdf = lcdf;
+    adjusted_lcdf[n] = 0;
+  } else if (max_strat == 2) {
+    // normalize
+    adjusted_lcdf = lcdf - lcdf[n];
+  } else {
+    reject("Unknown strategy to handle probability mass beyond the maximum value.");
+  }
+  return(adjusted_lcdf);
+}
+
+// compute discretised log scale pmf
+vector lcdf_to_log_prob(vector lcdf, int n) {
+  vector[n] lpmf; 
+  lpmf[1] = lcdf[1];
+  lpmf[2:n] = log_diff_exp(lcdf[2:n], lcdf[1:(n-1)]);
+  return(lpmf);
+}
+
+// compute discretised logit hazard
+vector lcdf_to_logit_hazard(vector lcdf, int n) {
+  vector[n] lhaz;
+  vector[n] lccdf;
+  lccdf = log1m_exp(lcdf);
+  lhaz[1] = lcdf[1];
+  lhaz[2:(n-1)] = log1m_exp(lccdf[2:(n-1)] - lccdf[1:(n-2)]);
+  lhaz[1:(n-1)] = lhaz[1:(n-1)] - log1m_exp(lhaz[1:(n-1)]);
+  lhaz[n] = positive_infinity();
```

Comment:
Maybe you can try `lhaz[n] = logit(1);` just for a test?

Reply:
yeah, I have tried a few variations of this and I get the issue in all of them (including converting everything back to the natural scale and doing it as you did originally). Really odd and not clear why. The only thing I have left to try I think is moving the `logit` out of the function and into the core stan code. Not at all clear why that would make a difference though.

Reply:
So I rolled back the hazard part to your version in https://github.com/epiforecasts/epinowcast/pull/140/commits/17378861a058f6c92babdbf6f1c42e7489f05b5a

This works as expected. In the next couple of commits will iterate on this to find what causes the infinite gradient.

Reply:
As soon as you switch the ccdf calculation to `exp(log1m_exp(lcdf))` then infinite gradients become an issue. Could be because `log1m_exp` is unstable when `lcdf` is near/equal to 1? (as results in `-Inf`)?

Reply:
https://github.com/epiforecasts/epinowcast/pull/140/commits/d992d5f19db4637c73ed88b603ebc3698398004b restricts the `ccdf` calculation to exclude the maximum (as known to be 1) and this now works as expected.

Reply:
The change to not calculate the maximum `lccdf` was suffiicent. https://github.com/epiforecasts/epinowcast/pull/140/commits/e6efb586329031fcd86c74f85b74cc77f3e0cfcb works as expected. This does make me concerned this may be unstable in settings where the maximum delay is much greater than the non-zero elements of the parametric delay.

Reply:
I constructed an extreme version of this scenario with no reporting beyond 10 days and a maximum delay of 20 days. It throws a few extra exceptions at initialisation but otherwise works well. 


```
# Load epinowcast and data.table
library(epinowcast)
library(data.table)

# Load and filter germany hospitalisations
nat_germany_hosp <- germany_covid19_hosp[location == "DE"][age_group %in% "00+"]
nat_germany_hosp <- enw_filter_report_dates(
  nat_germany_hosp,
  latest_date = "2021-10-01"
)

# Make sure observations are complete
nat_germany_hosp <- enw_complete_dates(
  nat_germany_hosp,
  by = c("location", "age_group")
)

# Exclude reported data beyond 10 days
nat_germany_hosp[report_date >= (reference_date + 10), confirm := NA]
nat_germany_hosp[,
 confirm := nafill(confirm, "locf"),
 by = c("reference_date")
]
# Make a retrospective dataset
retro_nat_germany <- enw_filter_report_dates(
  nat_germany_hosp,
  remove_days = 40
)
retro_nat_germany <- enw_filter_reference_dates(
  retro_nat_germany,
  include_days = 40
)

# Get latest observations for the same time period
latest_obs <- enw_latest_data(nat_germany_hosp)
latest_obs <- enw_filter_reference_dates(
  latest_obs,
  remove_days = 40, include_days = 20
)

# Preprocess observations
pobs <- enw_preprocess_data(retro_nat_germany, max_delay = 20)

# Reference date model
reference_module <- enw_reference(~1, data = pobs)

# Report date model
report_module <- enw_report(~ (1 | day_of_week), data = pobs)

# Compile nowcasting model
model <- enw_model(threads = TRUE)

# Fit nowcast model and produce a nowcast
# Note that we have reduced samples for this example to reduce runtimes
nowcast <- epinowcast(pobs,
  reference = reference_module,
  report = report_module,
  fit = enw_fit_opts(
    save_warmup = FALSE, pp = TRUE,
    chains = 2, threads_per_chain = 2,
    iter_warmup = 500, iter_sampling = 500
  ),
  model = model,
)

plot(nowcast, latest_obs)

plot(nowcast, type = "posterior") +
  facet_wrap(vars(reference_date), scale = "free")
```
[nowcast.pdf](https://github.com/epiforecasts/epinowcast/files/9198904/nowcast.pdf)
[plot.pdf](https://github.com/epiforecasts/epinowcast/files/9198905/plot.pdf)